code/renderers/vue3/src/render.ts

```diff
@@ -22,65 +30,128 @@ export const setup = (fn: (app: any) => void) => {
   if (existingApp && !forceRemount) {
-    updateArgs(existingApp.reactiveArgs, reactiveArgs);
+    updateContextDecorator(storyFn, storyContext);
+    updateArgs(existingApp.reactiveArgs, storyContext.args);
     return () => {
       teardown(existingApp.vueApp, canvasElement);
     };
   }
-
   if (existingApp && forceRemount) teardown(existingApp.vueApp, canvasElement);
 
-  const storybookApp = createApp({
-    render() {
-      const renderedElement: any = elementMap.get(canvasElement);
-      const current = renderedElement && renderedElement.template ? renderedElement : element;
-      map.set(canvasElement, { vueApp: storybookApp, reactiveArgs });
-      return h(current, reactiveArgs);
+  // create vue app for the story
+  const vueApp = createApp({
+    setup() {
+      storyContext.args = reactiveArgs;
+      reactiveState = reactive({ globals: storyContext.globals });
+      let rootElement: StoryFnVueReturnType = storyFn();
+
+      watch(
+        () => reactiveState.globals,
+        () => {
+          storyContext.globals = reactiveState.globals;
+          rootElement = storyFn();
+        }
+      );
+
+      return () => {
+        return h(rootElement, reactiveArgs);
+      };
+    },
+    mounted() {
+      map.set(canvasElement, {
+        vueApp,
+        reactiveArgs,
+      });
+    },
+    renderTracked(event) {
+      // console.log('vueApp renderTracked', event); // this works only in dev mode
+    },
+    renderTriggered(event) {
+      // console.log('vueApp renderTriggered', event); // this works only in dev mode
     },
   });
-
-  storybookApp.config.errorHandler = (e: unknown) => showException(e as Error);
-  setupFunction(storybookApp);
-  storybookApp.mount(canvasElement);
+  vueApp.config.errorHandler = (e: unknown) => showException(e as Error);
+  setupFunction(vueApp);
+  vueApp.mount(canvasElement);
 
   showMain();
   return () => {
-    teardown(storybookApp, canvasElement);
+    teardown(vueApp, canvasElement);
   };
 }
 
 /**
- * get the slots as functions to be rendered
- * @param props
+ * generate slots for default story without render function template
  * @param context
  */
 
-function getSlots(props: Args, context: StoryContext<VueRenderer, Args>) {
+function generateSlots(context: StoryContext<VueRenderer, Args>) {
   const { argTypes } = context;
-  const slots = Object.entries(props)
+  const slots = Object.entries(argTypes)
     .filter(([key, value]) => argTypes[key]?.table?.category === 'slots')
-    .map(([key, value]) => [key, typeof value === 'function' ? value : () => value]);
+    .map(([key, value]) => [
+      key,
+      () => {
+        const slotValue = context.args[key];
+        if (typeof slotValue === 'function' || isVNode(slotValue)) return h(slotValue);
+        if (Array.isArray(slotValue)) return slotValue.map((item: any) => h(item));
+        if (typeof slotValue === 'object') return JSON.stringify(slotValue);
```

Comment:
I don't really understand why we would need all of this, in the type in TS we only allow for:

* A `VNodeChild`
* or a function that accept props and returns a `VNodeChild`

If it is a function we can just return the value right? As the h function accepts that?
If it is an array, it must be an array of VNodeChild, and the h function also accepts that for child.
If it is an object, it must be a VNode, why should we stringify that?

It this is about doing something sensible when the user gives incorrect args, I would rather throw when the slot arg is not a Vnode or a function.

Reply:
Yes i got your point but dont forget the controls you can pass and object, rather we handle this case, for now we can just print , that why i stringify

Reply:
So for me the JSON like a fallback in case the object is not a VNode, there is no harm to have it .
1- VNode
2- Function Component 
3- Array of VNode  we need this 100% very usefull generating in UI , like list item , or carousel ...

Reply:
Actually, this doesn't take into account slots accepting inner props anymore:
<img width="1146" alt="image" src="https://user-images.githubusercontent.com/1035299/224734655-2b39b7b3-a92e-4a8f-8932-7156e9d94381.png">

I am not a fan of adding complexity here, it easily leads to more bugs. The h function is already powerful, I think the concern of storybook, should be to only pass the right args to slot function:
`h(Component, props, slot)`

And should not add any arbitrary mapping logic on top of the `h` function, IMO.

I think it is best if it works similar as if the user would write a custom render function based on `h`. For example:

```
export const Primary: Story = {
  render: () => {
    return h(
      BaseLayout,
      { otherProp: true },
      {
        header: ({ title }) =>
          h({
            components: { Button },
            template: `<Button :primary='true' label='${title}'></Button>`,
          }),
        default: [
          'some text',
          h(Button, { primary: true, label: 'default' }),
        ],
        footer: h(Button, { primary: true, label: 'footer' }),
      }
    );
  },
};
```

> 2- Function Component

I think the user should wrap themself the Function component in a `h` function.

> 3 Array of VNode we need this 100% very usefull generating in UI , like list item , or carousel ...

You can already pass array of VNode's to the slots in the `h` function, so we don't have to do anything there.

> 4 and you forgot the basic one which just a text or function returning text. before i go for complex type user may need just to pass a text

With `VNodeChild` I mean the following type def from the vue source:

```ts
type VNodeChild = VNodeChildAtom | VNodeArrayChildren;

type VNodeArrayChildren = Array<VNodeArrayChildren | VNodeChildAtom>;

type VNodeChildAtom = VNode | string | number | boolean | null | undefined | void;
```

So string, number boolean, null and undefined are covered there automatically by the `h` function. And arrays as well.

> Yes i got your point but dont forget the controls you can pass an object

This is also true for React, but we don't stringify children in React as well. Maybe at some point we could solve this more generally for all renderers.  For example, we should only allow the control of slots for any renderer to be updated with plain strings.

Reply:
> And should not add any arbitrary mapping logic on top of the `h` function, IMO.

i totally agree, i look at it now it is like doing extra mapping, it would work simple like this 
```ts 
 return typeof slotValue === 'function' ? h(slotValue) : slotValue;
```

if we ignore the JSON, as we said earlier we need to work on the controls

Reply:
> > And should not add any arbitrary mapping logic on top of the `h` function, IMO.
> 
> i totally agree, i look at it now it is like doing extra mapping, it would work simply like this
> 
> ```ts
>  return typeof slotValue === 'function' ? h(slotValue) : slotValue;
> ```
> 
> if we ignore the JSON, as we said earlier we need to work on the controls

Reply:
Actually, this still doesn't work with inner props, we probably should add a test case for that, as it easy to regress:

![image](https://user-images.githubusercontent.com/1035299/225042667-ec34980c-533e-41f6-91d6-e44a96bd3c5b.png)

See that my header is getting undefined, as the innerProps are not correctly passed.

It should be like this, to work correctly with innerProps:

```ts
  .map(([key, value]) => {
    return [key, context.args[key]];
  });
```

Or maybe better:

```ts
  .map(([key, value]) => {
    const slotValue = context.args[key];
    const normalized = typeof slotValue === 'function' ? slotValue : () => slotValue;
    return [key, normalized];
  });
```

As Vue gives performance warnings if slots are not specified as functions.